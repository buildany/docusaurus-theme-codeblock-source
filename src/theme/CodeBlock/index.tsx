import React from "react";
import DocusaurusCodeBlock from "../DocusaurusCodeBlock";
import { validUrl } from "../../utils";
import cx from "classnames";

import type { CodeBlockProps } from "../types";

const CodeBlock = ({
  title,
  sourceUrl,
  sourceClassName,
  ...props
}: CodeBlockProps) => {
  const url = validUrl(sourceUrl);
  const codeBlockTitle = url ? (
    <a
      data-testid="sourceCodeLink"
      href={url}
      target="_blank"
      className={cx("source-code-link", sourceClassName)}
    >
      {!!title ? title : url}
    </a>
  ) : (
    title
  );

  return <DocusaurusCodeBlock {...props} title={codeBlockTitle} />;
};

export default CodeBlock;
