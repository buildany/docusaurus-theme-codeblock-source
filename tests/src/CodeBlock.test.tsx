import React from "react";

import CodeBlock from "../../src/theme/CodeBlock";
import { render, waitFor, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

jest.mock(
  "../../src/theme/DocusaurusCodeBlock",
  () =>
    ({ className, title }: { className: string; title: string | Element }) =>
      (
        <div data-testid="docusaurusCodeBlock" className={className}>
          <div data-testid="title">{title}</div>
        </div>
      )
);
it("should render code block correctly without source code link", async () => {
  const title = "Code block title";
  const { asFragment } = render(
    <CodeBlock title={title}>
      <div id="codeBlockChildren" />
    </CodeBlock>
  );

  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => screen.getByTestId("docusaurusCodeBlock"));
  expect(screen.getByTestId("title").textContent).toBe(title);
});

it("should render code block correctly with source code link", async () => {
  const title = "Code block title";
  const sourceUrl = "https://example.code.org/file.js";
  const { asFragment } = render(
    <CodeBlock title={title} sourceUrl={sourceUrl}>
      <div id="codeBlockChildren" />
    </CodeBlock>
  );

  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => screen.getByTestId("docusaurusCodeBlock"));
  expect(screen.getByTestId("sourceCodeLink").textContent).toBe(title);
  expect(screen.getByTestId("sourceCodeLink")).toHaveAttribute(
    "href",
    sourceUrl
  );
  expect(screen.getByTestId("sourceCodeLink")).toHaveClass("source-code-link");
});

it("should render code block correctly with source code link if title is not specified", async () => {
  const sourceUrl = "https://example.code.org/file.js";
  const { asFragment } = render(
    <CodeBlock sourceUrl={sourceUrl}>
      <div id="codeBlockChildren" />
    </CodeBlock>
  );

  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => screen.getByTestId("docusaurusCodeBlock"));
  expect(screen.getByTestId("sourceCodeLink").textContent).toBe(sourceUrl);
  expect(screen.getByTestId("sourceCodeLink")).toHaveAttribute(
    "href",
    sourceUrl
  );
  expect(screen.getByTestId("sourceCodeLink")).toHaveClass("source-code-link");
});

it("should render code block correctly with source code link and custom class name", async () => {
  const title = "Code block title";
  const sourceUrl = "https://example.code.org/file.js";
  const sourceClassName = "source-link-custom-class";
  const { asFragment } = render(
    <CodeBlock
      title={title}
      sourceUrl={sourceUrl}
      sourceClassName={sourceClassName}
    >
      <div id="codeBlockChildren" />
    </CodeBlock>
  );

  expect(asFragment()).toMatchSnapshot();
  await waitFor(() => screen.getByTestId("docusaurusCodeBlock"));
  expect(screen.getByTestId("sourceCodeLink").textContent).toBe(title);
  expect(screen.getByTestId("sourceCodeLink")).toHaveAttribute(
    "href",
    sourceUrl
  );
  expect(screen.getByTestId("sourceCodeLink")).toHaveClass("source-code-link");
  expect(screen.getByTestId("sourceCodeLink")).toHaveClass(sourceClassName);
});
