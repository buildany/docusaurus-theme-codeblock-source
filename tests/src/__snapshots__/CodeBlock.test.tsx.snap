// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render code block correctly with source code link 1`] = `
<DocumentFragment>
  <div
    data-testid="docusaurusCodeBlock"
  >
    <div
      data-testid="title"
    >
      <a
        class="source-code-link"
        data-testid="sourceCodeLink"
        href="https://example.code.org/file.js"
        target="_blank"
      >
        Code block title
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render code block correctly with source code link and custom class name 1`] = `
<DocumentFragment>
  <div
    data-testid="docusaurusCodeBlock"
  >
    <div
      data-testid="title"
    >
      <a
        class="source-code-link source-link-custom-class"
        data-testid="sourceCodeLink"
        href="https://example.code.org/file.js"
        target="_blank"
      >
        Code block title
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render code block correctly with source code link if title is not specified 1`] = `
<DocumentFragment>
  <div
    data-testid="docusaurusCodeBlock"
  >
    <div
      data-testid="title"
    >
      <a
        class="source-code-link"
        data-testid="sourceCodeLink"
        href="https://example.code.org/file.js"
        target="_blank"
      >
        https://example.code.org/file.js
      </a>
    </div>
  </div>
</DocumentFragment>
`;

exports[`should render code block correctly without source code link 1`] = `
<DocumentFragment>
  <div
    data-testid="docusaurusCodeBlock"
  >
    <div
      data-testid="title"
    >
      Code block title
    </div>
  </div>
</DocumentFragment>
`;
